inbox/interruptible_threading.py

```diff
@@ -0,0 +1,263 @@
+            raise NotImplementedError()
+
+    def kill(self, block: bool = True) -> None:
+        """
+        Kill the thread.
+
+        If block is True, wait until the thread is ready.
+        """
+        self.__should_be_killed.set()
+        if block:
+            self.join()
+
+    def _check_interrupted(self) -> None:
+        """
+        Internally check if the thread should be interrupted.
+
+        Don't use this directly instead use the public
+        `check_interrupted` function below.
+        """
+        if self.__should_be_killed.wait(0.01):
+            raise InterruptibleThreadExit()
+
+        if (
+            self._timeout_deadline is not None
+            and time.monotonic() >= self._timeout_deadline
+        ):
+            raise InterruptibleThreadTimeout()
+
+
+P = ParamSpec("P")
+T = TypeVar("T")
+
+
+def _interruptible(
+    blocking_function: Callable[P, T]
+) -> Callable[[Callable[Concatenate[InterruptibleThread, P], T]], Callable[P, T]]:
+    """
+    If the current thread is interruptible run interruptible version of
+    the blocking function. Otherwise fallback to original implementation.
+    """
+
+    def decorator(
+        interruptible_function: Callable[Concatenate[InterruptibleThread, P], T]
+    ) -> Callable[P, T]:
+        def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
+            current_thread = threading.current_thread()
+            if not isinstance(current_thread, InterruptibleThread):
+                return blocking_function(*args)
+
+            return interruptible_function(current_thread, *args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+# Time to wait between checking if the thread should be interrupted
+# when using interruptible versions of blocking functions.
+CHECK_INTERRUPTED_TIMEOUT = 0.2
+
+
+@_interruptible(time.sleep)
+def sleep(current_thread: InterruptibleThread, /, seconds: float) -> None:
+    """
+    Interruptible version of time.sleep.
+    """
+    start = time.monotonic()
+    while time.monotonic() - start < seconds:
+        time.sleep(CHECK_INTERRUPTED_TIMEOUT)
+        current_thread._check_interrupted()
+
+
+@_interruptible(queue.Queue.get)
+def queue_get(
+    current_thread: InterruptibleThread,
+    /,
+    self: "queue.Queue[queue._T]",
+    block: bool = True,
+    timeout: "float | None" = None,
+) -> "queue._T":
+    """
+    Interruptible version of queue.Queue.get.
+    """
+    if not block:
+        return self.get(block=False)
+
+    if timeout is not None:
+        raise NotImplementedError("timeout is not supported.")
+
+    while True:
+        try:
+            return self.get(timeout=CHECK_INTERRUPTED_TIMEOUT)
+        except queue.Empty:
+            current_thread._check_interrupted()
+
+
+@_interruptible(lambda: None)
+def check_interrupted(current_thread: InterruptibleThread, /) -> None:
+    """
+    Check if the current thread is interrupted.
+    """
+    return current_thread._check_interrupted()
```

Comment:
Why return a method call that returns `None` and only communicates by raising an exception?

Reply:
This is exactly how gevent works.

Reply:
This is what https://greenlet.readthedocs.io/en/latest/api.html#greenlet.GreenletExit does. In gevent if you call `.kill()` on a greenlet it will raise `GreenletExit` next time it uses one of the gevent APIs or monkeypatched stdlib.

Reply:
@mrhiggi-close Also asyncio works exactly the same way in this regard, it raises an exception https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.CancelledError if you `.cancel()` a running task.

Reply:
Control flow with exceptions is a standard thing with concurrency libraries. I will cover this in S&T.